### PR TITLE
Various updates, fixes and additions

### PR DESCRIPTION
**Pull request by adamzero1**
- feat: box-project version updated
- fixed: unable to build project phar
- feat: added the ability to specify multiple code parse types, php/di.xml/composer.json/module.xml 
  These can be specified as desired, all `--all` can be used
- feat: help text, now shows a version (just for ease in future should there be more updates)
- docs: updated docs with descriptions/commands on how to build and test
- feat: added dockerfile, so builds can be done in a common environment
- refactor: have refactored some of the clases into common directories, such as `code parsers` and `code providers`
- feat: have added the option to allow for a json output (so the output can be easily parsed by other scripts)
- feat: have added the option to "enable module name" to add the referenced classes module name in the output
- feat: have added the option to "enable source file" to add the referenced classes source file path in the output
- feat: code providers can now be passed a list of valid patterns, this has two bonuses: 
  - they can return multiple file types
  - they only return file types we want
- feat: added some utility scripts to the bin directory, for ease of dev / testing / building
- fixed: issue where references would sometimes be returned as a standard array and sometimes as an assoc array, this is now always a standard array
- feat: added code parser type `composer.json`
- feat: added code parser type `module.xml`

  
All changes should be backwards compatible.

### CODE REVIEW

**Review by Vinai — CHANGES_REQUESTED** [review]
This is great! The only thing that bugs me a little bit is some of the formatting.
So for example, for [functions](https://www.php-fig.org/psr/psr-12/#44-methods-and-functions), instead of `printHelpEndExit(){` I am used to the PSR-2 - or,12 nowadays, I guess - style
```php
printHelpEndExit()
{
```

And for [conditionals](https://www.php-fig.org/psr/psr-12/#51-if-elseif-else) like `if(count($argv) > 1){` a space between the parens and braces and such

```php
if (count($argv) > 1) {
```

I tried to be bigger and let it go but I couldn't ;)
Happy to apply the changes myself after merging if you don't like it.

**Comment by adamzero1** [comment]
No worries, I've added gitbub workflow file to the project that should check for PSR12.
Though it doesn't seem to be running, not sure if this is because it's a fork and I don't have permissions or something.

Let me know if you need / want anything else

**Review by Vinai — APPROVED** [review]
LGTM

**Review by fballiano — APPROVED** [review]
since @Vinai is the original author and approved this PR, I'm approving to allow merging

**Comment by fballiano** [comment]
@adamzero1 @Vinai @sprankhub @rhoerr do you know who has write privileges to this repo? we can't merge this PR.

File: bin/build-docker-image [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -xe
+
+docker build --build-arg UUID=$(id -u) --build-arg GUID=$(id -g) ./build --tag php-dependency-list-env:1.0.0
```
Comment (fballiano) [inline bin/build-docker-image]:
```suggestion
docker build --build-arg UUID=$(id -u) --build-arg GUID=$(id -g) ./build --tag php-dependency-list-env:1.0.0

```

File: bin/build-phar [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -xe
+
+DOCKER_IMAGE=php-dependency-list-env:1.0.0
+
+if [[ "$(docker images -q ${DOCKER_IMAGE} 2> /dev/null)" == "" ]]; then
+  ./bin/build-docker-image
+fi
+
+
+docker run -t --volume $(pwd):/usr/src ${DOCKER_IMAGE} bash -c "make; exit $?"
```
Comment (fballiano) [inline bin/build-phar]:
```suggestion
docker run -t --volume $(pwd):/usr/src ${DOCKER_IMAGE} bash -c "make; exit $?"

```

File: box.json.dist [hunk 3]
```diff
@@ -1,7 +1,5 @@
 {
-  "files": [
-    "bin/php-classes.php"
-  ],
-  "force-autodiscovery": true,
-  "output": "php-classes.phar"
+    "chmod": "0700",
+    "main": "bin/php-classes.php",
+    "output": "php-classes.phar"
 }
```
Comment (fballiano) [inline box.json.dist]:
```suggestion
}

```

File: build/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	&& docker-php-ext-install -j$(nproc) zip \
+    && docker-php-ext-configure sockets \
+	&& docker-php-ext-install -j$(nproc) sockets
+
+# create user to run as
+RUN groupadd --gid ${GUID} --non-unique mage \
+    && useradd --gid mage --shell /bin/bash --uid ${UUID} --no-user-group --non-unique mage
+
+WORKDIR /usr/src
+USER mage
```
Comment (fballiano) [inline build/Dockerfile]:
```suggestion
USER mage

```

**Review by fballiano — APPROVED** [review]
minor "empty line at end of files" suggestion

**Comment by sprankhub** [comment]
> @adamzero1 @Vinai @sprankhub @rhoerr do you know who has write privileges to this repo? we can't merge this PR.

Definitely `mage-os-ci`. Besides that, no idea from the top of my head. It seems to not be included in https://github.com/mage-os/terraform/blob/main/variables.tf, which is probably an oversight. I added it to https://github.com/mage-os/terraform/issues/14 now.

**Comment by fballiano** [comment]
@mage-os-ci @mage-os/infrastructure could somebody take a look at this?

**Pull request merged**